### ISSUE TITLE
Erratic memory consumption

### ISSUE THREAD

**Issue by neoxic** [LuaJIT/LuaJIT#622]
Hello,

Attention! Please **preserve binary names, Lua file names and Lua code** as they appear here since the bug seems to be very flaky.

Steps to build LuaJIT:
```
$ git clone https://github.com/LuaJIT/LuaJIT.git
$ cd LuaJIT
$ make amalg XCFLAGS="-DLUA_USE_APICHECK -DLUA_USE_ASSERT -Og -g"
$ make install
```
Please consider the following test cases.

test1.lua
---
```Lua
local T = {F = function () end}

local function copy(x, r, p)
    if type(x) ~= 'table' then
        return x
    end
    r = r or {}
    p = p or {}
    local t = r[x]
    if t then
        return t
    end
    t = {}
    r[x] = t
    local iterator = pairs -- Don't ask!
    for k, v in iterator(x) do
        table.insert(p, k)
        t[k] = copy(v, r)
        table.remove(p)
    end
    return t
end

while true do
    copy(T)
end
```
```
$ luajit-2.1.0-beta3 test1.lua
PANIC: unprotected error in call to Lua API (not enough memory)
```

test2.lua
---
```Lua
local T = {F = function () end}

local function copy(x, r, p)
    if type(x) ~= 'table' then
        return x
    end
    r = r or {}
    p = p or {}
    local t = r[x]
    if t then
        return t
    end
    t = {}
    r[x] = t
    local iterator = pairs -- Don't ask!
    for k, v in iterator(x) do
        table.insert(p, k)
        t[k] = copy(v, r)
        table.remove(p)
    end
    return t
end

local i = 0
while true do
    i = i + 1
    if i == 1e6 then
        i = 0
        print(('Memory usage: %.3f Mb'):format(collectgarbage('count') / 1024))
    end
    copy(T)
end
```
```
$ luajit-2.1.0-beta3 test2.lua
Memory usage: 297.587 Mb
Memory usage: 282.935 Mb
Memory usage: 283.656 Mb
Memory usage: 283.619 Mb
Memory usage: 283.621 Mb
Memory usage: 283.621 Mb
Memory usage: 283.621 Mb
^C
```

**Comment by neoxic** [comment]
The above test cases don't seem to affect a 32-bit version of LuaJIT without extra fiddling.

Rebuilding LuaJIT as 32-bit:
```
$ make clean
$ make amalg XCFLAGS="-DLUA_USE_APICHECK -DLUA_USE_ASSERT -Og -g" CC="gcc -m32"
$ make install
```

test3.lua
---
```Lua
local T = {F = function () end}

local function copy(x, r, p)
    if type(x) ~= 'table' then
        return x
    end
    r = r or {}
    p = p or {}
    local t = r[x]
    if t then
        return t
    end
    t = {}
    r[x] = t
    local iterator = pairs -- Don't ask!
    for k, v in iterator(x) do
        table.insert(p, k)
        t[k] = copy(v, r)
        table.remove(p)
    end
    return t
end

local table_insert = table.insert
local i = 0
while true do
    i = i + 1
    if i == 1e6 then
        i = 0
        function table.insert(...)
            return table_insert(...)
        end
        print(('Memory usage: %.3f Mb'):format(collectgarbage('count') / 1024))
    end
    copy(T)
end
```
```
$ luajit-2.1.0-beta3 test3.lua
Memory usage: 0.068 Mb
Memory usage: 0.071 Mb
Memory usage: 206.067 Mb
Memory usage: 191.376 Mb
Memory usage: 192.419 Mb
Memory usage: 192.345 Mb
Memory usage: 192.350 Mb
Memory usage: 192.350 Mb
Memory usage: 192.350 Mb
Memory usage: 192.350 Mb
Memory usage: 192.350 Mb
^C
```

**Comment by neoxic** [comment]
For the reference, the same tests on vanilla Lua 5.1:
```
$ lua test1.lua
^C

$ lua test2.lua
Memory usage: 0.047 Mb
Memory usage: 0.032 Mb
Memory usage: 0.053 Mb
Memory usage: 0.038 Mb
Memory usage: 0.060 Mb
Memory usage: 0.045 Mb
Memory usage: 0.053 Mb
^C

$ lua test3.lua
Memory usage: 0.032 Mb
Memory usage: 0.062 Mb
Memory usage: 0.056 Mb
Memory usage: 0.049 Mb
Memory usage: 0.042 Mb
Memory usage: 0.036 Mb
Memory usage: 0.066 Mb
^C
```

LuaJIT 2.0.5 seems to be unaffected too.

**Comment by ExtReMLapin** [comment]
Can confirm, bug doesn't happen on Torch ` 2.1.0-beta2` build.
However it generates a crazy ton of traces.

On HEAD, it generates only very few traces but had very high only with jit compiler on.


**JIT Compiler OFF : **
```txt
PS C:\Users\pierr\Desktop> luajit -joff .\script2.lua
Memory usage: 0.040 Mb
Memory usage: 0.035 Mb
Memory usage: 0.063 Mb
Memory usage: 0.058 Mb
Memory usage: 0.053 Mb
Memory usage: 0.048 Mb
Memory usage: 0.043 Mb
Memory usage: 0.037 Mb
...
```

With JIT compiler on and with trace report : 

```txt
PS C:\Users\pierr\Desktop> luajit -jv .\script2.lua
[TRACE --- script2.lua:16 -- leaving loop in root trace at script2.lua:21]
[TRACE   1 script2.lua:16 stitch next]
[TRACE   2 (1/stitch) script2.lua:16 stitch next]
[TRACE   3 (2/stitch) script2.lua:16 -> 1]
[TRACE   4 script2.lua:26 stitch next]
[TRACE   5 (4/stitch) script2.lua:16 -> 1]
Memory usage: 0.092 Mb
[TRACE   6 (1/0) script2.lua:17 stitch next]
[TRACE   7 (6/stitch) script2.lua:16 -> 4]
[TRACE   8 script2.lua:3 stitch next]
[TRACE   9 (8/stitch) script2.lua:16 -> 1]
Memory usage: 0.116 Mb
[TRACE  10 (6/0) script2.lua:17 -> 8]
[TRACE  11 (8/0) script2.lua:4 stitch next]
[TRACE  12 (11/stitch) script2.lua:16 -> 4]
Memory usage: 206.114 Mb
[TRACE  13 (10/0) script2.lua:17 -> 8]
Memory usage: 191.396 Mb
Memory usage: 192.440 Mb
Memory usage: 192.367 Mb
Memory usage: 192.372 Mb
Memory usage: 192.372 Mb
Memory usage: 192.372 Mb
```

**Comment by MikePall** [comment]
OK, I now have a good idea what's happening:

Due to trace stitching a couple traces jump to each other forming a loop. The trace with the allocations actually does drive the GC forward. And exit 0 of this trace is supposed to return back to the interpreter occasionally to perform the atomic GC phase, because that cannot be run on-trace. But exit 0 is extended into a side-trace (which doesn't exit) and loops back without an exit to the interpreter. Thus the GC is stuck just before the atomic phase and memory grows unboundedly.

The variants which print the memory usage every 1e6 iterations _do_ eventually run the atomic phase. That's why they keep on printing the same high numbers (just before the sweep) instead of blowing up.

I think the trace stitching optimization is not strictly necessary for this to happen. Either `-O-loop` or some clever testcase could trigger it, too. But I still need to come up with a smaller and more reliable testcase.

Anyway, I'm still not sure how to resolve this issue. Either force a GC check in the side trace at exit 0 (but that one could share the same fate) or separate the GC check exit (major rewrite) or something else I haven't though of, yet. Tough issue.

**Comment by neoxic** [comment]
Since you're baffled, I guess LuaJIT 2.0 has the same problem, right?

**Comment by MikePall** [comment]
Fixed for both v2.0 and v2.1 by not patching the exit branch for the GC check.

I still haven't come up with a reliable testcase or one that works with v2.0. But I've carefully verified the branches are not patched anymore for all archs.
